### ISSUE TITLE
Path -> Segment -> Relationship start/end node ID inverted

### ISSUE THREAD

**Issue by Meatballs1** [neo4jrb/neo4j-ruby-driver#51]
I get the following results from running a query against my DB, from the Code > Response output from the Neo4j browser

```
[
  {
    "keys": [
      "path"
    ],
    "length": 1,
    "_fields": [
      {
        "start": {
          "identity": {
            "low": 545,
            "high": 0
          },
          "labels": [
            "group"
          ],
          "properties": {
            "name": "node_545",
          }
        },
        "end": {
          "identity": {
            "low": 545,
            "high": 0
          },
          "labels": [
            "group"
          ],
          "properties": {
            "name": "node_545",
		  }
        },
        "segments": [],
        "length": 0
      }
    ],
    "_fieldLookup": {
      "path": 0
    }
  },
  {
    "keys": [
      "path"
    ],
    "length": 1,
    "_fields": [
      {
        "start": {
          "identity": {
            "low": 545,
            "high": 0
          },
          "labels": [
            "group"
          ],
          "properties": {
            "name": "node_545",
          }
        },
        "end": {
          "identity": {
            "low": 197,
            "high": 0
          },
          "labels": [
            "container"
          ],
          "properties": {
            "name": "node_197",
          }
        },
        "segments": [
          {
            "start": {
              "identity": {
                "low": 545,
                "high": 0
              },
              "labels": [
                "group"
              ],
              "properties": {
                "name": "node_545",
              }
            },
            "relationship": {
              "identity": {
                "low": 1049,
                "high": 0
              },
              "start": {
                "low": 197,
                "high": 0
              },
              "end": {
                "low": 545,
                "high": 0
              },
              "type": "TEST_RELATIONSHIP",
              "properties": {
                "cost": 0.7
              }
            },
            "end": {
              "identity": {
                "low": 197,
                "high": 0
              },
              "labels": [
                "ObjectDN",
                "container"
              ],
              "properties": {
                "name": "node_197",
              }
            }
          }
        ],
        "length": 1
      }
    ],
    "_fieldLookup": {
      "path": 0
    }
  }
]
```

First Path fields shows Start Node 545 end node 545.
Second Path shows a start node 545 and end node 197
	Segment in this path shows start node 545 and end node 197
		Relationship in this path shows start node 197 and end node 545

This suggests that I should have a relationship like:
 
`197-[TEST_RELATIONSHIP]->545`

However, when I process this result with the ruby driver I get:

```
statement_result.each do |record|
      path = record.to_h[:path]
      path.each do |segment|
	    r = segment.relationship
        logger.info("#{r.start_node_id}-[#{r.type}]->#{r.end_node_id}")
      end
end
```

`545-[TEST_RELATIONSHIP]->197`

It appears the relationship object is being populated with the start/end node from the path or segment rather than the relationship object?

If the relationship filter is the other way `>` it seems fine.

**Comment by Meatballs1** [comment]
Test case using the `:play movie-graph` 

```
require 'active_support/logger'
require 'neo4j_ruby_driver'

query = """MATCH (obj:Movie {title:'The Birdcage'})
CALL apoc.path.expandConfig(
  obj,
  {minLevel: -1,
  maxLevel: 2,
  bfs: true,
  relationshipFilter: '<REVIEWED|<ACTED_IN|<FOLLOWS',
  filterStartNode: false,
  limit: -1,
  uniqueness: 'NODE_GLOBAL',
  optional: true}) yield path
return path"

Neo4j::Driver::GraphDatabase.driver("bolt://172.17.0.4:7687", Neo4j::Driver::AuthTokens.none,  encryption: false) do |driver|
  driver.session do |session|
    session.read_transaction do |tx|
      result = tx.run(query, nil)
      result.each do |record|
        path = record.to_h[:path]
        puts ("#{path.start_node.id} - #{path.start_node.properties[:title]}")
        path.each do |segment|
          r = segment.relationship
          puts ("#{r.start_node_id}-[#{r.type}]->#{r.end_node_id}")
        end
      end
    end
  end
end
```

We get the following result from Neo4j Browser:

![Screenshot from 2020-10-07 14^%49^%42](https://user-images.githubusercontent.com/1854557/95339874-737e8280-08ac-11eb-827b-d2a420c69bf7.png)

Output from script shows most relationships are going from the birdcage movie outwards rather than inwards.

```
4216 - The Birdcage
4216 - The Birdcage
4216-[REVIEWED]->4290
4216 - The Birdcage
4216-[ACTED_IN]->4181
4216 - The Birdcage
4216-[ACTED_IN]->4201
4216 - The Birdcage
4216-[ACTED_IN]->4210
4216 - The Birdcage
4216-[REVIEWED]->4291
4291-[FOLLOWS]->4290
4216 - The Birdcage
4216-[REVIEWED]->4289
4289-[FOLLOWS]->4290
```

**Comment by Meatballs1** [comment]
From the Java driver:

>      * A segment combines a relationship in a path with a start and end node that describe the traversal direction
>     * for that relationship. This exists because the relationship has a direction between the two nodes that is
  >   * separate and potentially different from the direction of the path.

**Comment by Meatballs1** [comment]
The logic all seems the same as https://github.com/neo4j/neo4j-java-driver/blob/b20a89bed61be69ffbc6d6149ea2277ef4f789d4/driver/src/main/java/org/neo4j/driver/internal/messaging/common/CommonValueUnpacker.java#L299 but I cant confirm if the Java driver has the same problem

**Comment by Meatballs1** [comment]
Logic seems fine, however `rel_idx` doesn't appear to ever be negative

**Comment by Meatballs1** [comment]
I popped in the rel_idx into the relationship properties for some cheap debugging:

`rel.properties[:rel_idx] = rel_idx`

and test case now shows:

```
4216 - The Birdcage
4216 - The Birdcage
4216-[REVIEWED]->4290 ({:summary=>"Slapstick redeemed only by the Robin Williams and Gene Hackman's stellar performances", :rating=>45, :rel_idx=>1})
4216 - The Birdcage
4216-[ACTED_IN]->4181 ({:roles=>["Armand Goldman"], :rel_idx=>1})
4216 - The Birdcage
4216-[ACTED_IN]->4201 ({:roles=>["Albert Goldman"], :rel_idx=>1})
4216 - The Birdcage
4216-[ACTED_IN]->4210 ({:roles=>["Sen. Kevin Keeley"], :rel_idx=>1})
4216 - The Birdcage
4216-[REVIEWED]->4291 ({:summary=>"Slapstick redeemed only by the Robin Williams and Gene Hackman's stellar performances", :rating=>45, :rel_idx=>1})
4291-[FOLLOWS]->4290 ({:rel_idx=>2})
4216 - The Birdcage
4216-[REVIEWED]->4289 ({:summary=>"Slapstick redeemed only by the Robin Williams and Gene Hackman's stellar performances", :rating=>45, :rel_idx=>1})
4289-[FOLLOWS]->4290 ({:rel_idx=>2})
```

**Comment by klobuczek** [comment]
@Meatballs1 could you run your case with the java driver or if you prefer with the neo4j-java-driver gem on JRuby. If java has the same problem then the ruby driver is consistent with that because I used the same algorithm.

**Comment by Meatballs1** [comment]
Looks fine with jruby:

```
$ jruby /opt/ad-control-web/test.rb 
Oct 07, 2020 4:44:22 PM org.neo4j.driver.internal.logging.JULogger info
INFO: Direct driver instance 660915526 created for server address 172.17.0.4:7687
4216 - The Birdcage
4216 - The Birdcage
4290-[REVIEWED]->4216 ({:summary=>"Slapstick redeemed only by the Robin Williams and Gene Hackman's stellar performances", :rating=>45})
4216 - The Birdcage
4181-[ACTED_IN]->4216 ({:roles=>["Armand Goldman"]})
4216 - The Birdcage
4201-[ACTED_IN]->4216 ({:roles=>["Albert Goldman"]})
4216 - The Birdcage
4210-[ACTED_IN]->4216 ({:roles=>["Sen. Kevin Keeley"]})
4216 - The Birdcage
4290-[REVIEWED]->4216 ({:summary=>"Slapstick redeemed only by the Robin Williams and Gene Hackman's stellar performances", :rating=>45})
4291-[FOLLOWS]->4290 ({})
4216 - The Birdcage
4290-[REVIEWED]->4216 ({:summary=>"Slapstick redeemed only by the Robin Williams and Gene Hackman's stellar performances", :rating=>45})
4289-[FOLLOWS]->4290 ({})
Oct 07, 2020 4:44:22 PM org.neo4j.driver.internal.logging.JULogger info
INFO: Closing driver instance 660915526
Oct 07, 2020 4:44:22 PM org.neo4j.driver.internal.logging.JULogger info
INFO: Closing connection pool towards 172.17.0.4:7687
```
Compared to:

```
$ ruby /opt/ad-control-web/test.rb 
4216 - The Birdcage
4216 - The Birdcage
4216-[REVIEWED]->4290 ({:summary=>"Slapstick redeemed only by the Robin Williams and Gene Hackman's stellar performances", :rating=>45, :rel_idx=>1})
4216 - The Birdcage
4216-[ACTED_IN]->4181 ({:roles=>["Armand Goldman"], :rel_idx=>1})
4216 - The Birdcage
4216-[ACTED_IN]->4201 ({:roles=>["Albert Goldman"], :rel_idx=>1})
4216 - The Birdcage
4216-[ACTED_IN]->4210 ({:roles=>["Sen. Kevin Keeley"], :rel_idx=>1})
4216 - The Birdcage
4216-[REVIEWED]->4291 ({:summary=>"Slapstick redeemed only by the Robin Williams and Gene Hackman's stellar performances", :rating=>45, :rel_idx=>1})
4291-[FOLLOWS]->4290 ({:rel_idx=>2})
4216 - The Birdcage
4216-[REVIEWED]->4289 ({:summary=>"Slapstick redeemed only by the Robin Williams and Gene Hackman's stellar performances", :rating=>45, :rel_idx=>1})
4289-[FOLLOWS]->4290 ({:rel_idx=>2})
```

I suspect the conversion of integers from seabolt to ruby maybe coming unsigned?

**Comment by klobuczek** [comment]
could you try to find a minimal example without the movie db and apoc

**Comment by Meatballs1** [comment]
```
MERGE (t:TARGET)
WITH t
MERGE (i:IN)
WITH i, t
MERGE (i)-[:IN]->(t)
```

```
require 'active_support/logger'
require 'neo4j_ruby_driver'

query = "MATCH p=(t:TARGET)<-[*]-() RETURN p AS path"

Neo4j::Driver::GraphDatabase.driver("bolt://172.17.0.4:7687", Neo4j::Driver::AuthTokens.none,  encryption: false) do |driver|
  driver.session do |session|
    session.read_transaction do |tx|
      result = tx.run(query, nil)
      result.each do |record|
        path = record.to_h[:path]
        path.each do |segment|
          r = segment.relationship
	  puts ("#{r.start_node_id}-[#{r.type}]->#{r.end_node_id} (#{r.properties})")
        end
      end
    end
  end
end
```

Results:

```
$ ruby test.rb 
4121-[IN]->4122 ({})
```

```
$ jruby test.rb 
4122-[IN]->4121 ({})
```

**Comment by klobuczek** [comment]
FIxed and released 1.7.2. 
This was the culprit: https://github.com/neo4jrb/neo4j-ruby-driver/commit/c5724b48cf596a8f671e2d9fcc60686707322757#diff-4e515074cec0327c930da758d5bfa3b1R16
Thanks for the report and the other PR.

**Comment by Meatballs1** [comment]
Ah, good spot :)
